### PR TITLE
added MAb related environments, added more controllers, code structure changed for better interactions between controllers and env, added documents.

### PR DESCRIPTION
**Pull request by mohan-zhang-quartic**

### CODE REVIEW

File: quarticgym/envs/beerfmtenv.py [hunk 1]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
-            reward = 200
-        elif done:  # reaches time limit but reaction has not finished
-            reward = -200
-        else:
-            reward = -1  # we want the simulation to end fast
-        self.total_reward += reward
+        observation = [X_A, X_L, X_D, S, EtOH, DY, EA, self.step_count + 1]
 
+        observation = np.array(observation, dtype=self.np_dtype)
+        # compute reward
```
Comment (Sile-Tao) [inline quarticgym/envs/beerfmtenv.py:255]: maybe these 2 comments can be removed

**Review by Sile-Tao — COMMENTED** [review]
They look good to me! I left some comments and I will continue the review tomorrow

File: quarticgym/envs/beerfmtenv.py [hunk 2]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
         The only action/input is temperature.
         The observations are X_A, X_L, X_D, S, EtOH, DY, EA, time
         """
+        
+        # define arguments
+        self.step_count = 0
+        self.total_reward = 0
+        self.done = False
         self.dense_reward = dense_reward
+        self.normalize = normalize  # whether we want to normalize the observation and action to be in between -1 and 1. This is common in most of RL algorithms
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:82]: sorry those are leftover comments from QuarticGymEnvBase. gonna remove them.

File: quarticgym/envs/beerfmtenv.py [hunk 3]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
 
 
 class BeerFMTEnvGym(QuarticGymEnvBase):
+    def __init__(self, dense_reward=True, normalize=True, debug_mode=False, action_dim=1, observation_dim=8,
+                 reward_function=None, done_calculator=None, max_observations=[15, 15, 15, 150, 150, 10, 10, MAX_STEPS],
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:66]: yes, and I believe max_observations/min_observations, max_actions/min_actions are common in reinforcement learning community, so I name them as they are, like here https://github.com/openai/gym/blob/master/gym/wrappers/rescale_action.py

File: quarticgym/envs/beerfmtenv.py [hunk 4]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
+                                                shape=(self.observation_dim,))
+            self.action_space = spaces.Box(low=self.min_actions, high=self.max_actions, shape=(self.action_dim,))
+        
         self.res_forplot = []  # for plotting purposes
+    
+    def reward_function_standard(self, previous_observation, action, current_observation, reward=None):
+        """the s, a, r, s, a calculation.
+
+        Args:
+            previous_observation ([type]): This is denormalized observation, as usual.
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:125]: updated.

File: quarticgym/envs/beerfmtenv.py [hunk 5]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
+            current_observation ([type]): This is denormalized observation, as usual.
+
+        Returns:
+            [float]: reward.
+        """
+        #
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:131]: updated.

File: quarticgym/envs/beerfmtenv.py [hunk 6]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
+        # 
+        if reward is not None:
+            return reward
+        elif self.observation_beyond_box(current_observation):
+            return self.error_reward
+
+        done, done_info = self.done_calculator(current_observation, self.step_count, reward)
+        early_submission = done_info.get('early_submission', False)
+        if early_submission:
+            reward = -self.error_reward # early submission is rewarded.
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:140]: more comments added

File: quarticgym/envs/beerfmtenv.py [hunk 7]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
         self.total_reward = 0
         self.done = False
-        observation = BEER_init
+
+        if initial_state is not None:
+            initial_state = np.array(initial_state, dtype=self.np_dtype)
+            observation = initial_state
+            self.init_observation = initial_state
+        else:
+            observation = self.sample_initial_state()
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:219]: again this is following the gym name design "observation_space" so I use singular here... I know that they are interchangably using singular and plurals, but it is reasonable for:
an observation element -> an array of observations
observation at time t -> observations for the whole trajectory

File: quarticgym/envs/beerfmtenv.py [hunk 8]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
+        """
+        self.step_count = 0
         self.total_reward = 0
         self.done = False
-        observation = BEER_init
+
+        if initial_state is not None:
+            initial_state = np.array(initial_state, dtype=self.np_dtype)
+            observation = initial_state
+            self.init_observation = initial_state
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:217]: same as above

File: quarticgym/envs/beerfmtenv.py [hunk 9]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
+        done_info = {"terminal": False, "timeout": False}
+        action = np.array(action, dtype=self.np_dtype)
+        normalize = self.normalize if normalize is None else normalize
+        if normalize:
             action, _, _ = denormalize_spaces(action, self.max_actions, self.min_actions)
-        t = np.arange(0 + self.time, 1 + self.time, 0.01)
-        X_A, X_L, X_D, S, EtOH, DY, EA, _ = self.prev_denormalized_observation
+
+        # TOMODIFY: proceed your environment here and collect the observation.
+        t = np.arange(0 + self.step_count, 1 + self.step_count, 0.01)
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:247]: to remind myself it is calculating the start and plotting end for ode solver. I will name 0.01 to interval

File: quarticgym/envs/beerfmtenv.py [hunk 10]
```diff
@@ -66,86 +62,217 @@ def beer_ode(points, t, sets):
...
+        X_A, X_L, X_D, S, EtOH, DY, EA, _ = self.previous_observation
         sol = odeint(beer_ode, (X_A, X_L, X_D, S, EtOH, DY, EA), t, args=([INIT_SUGAR, action[0] + 273.15],))
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:249]: sure!

[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:249]: action's length is 1.It contains only temp. I just take the float item out of the index here, so I think this way it keeps things simple

File: quarticgym/envs/beerfmtenv.py (hunk 1 above)
Comment (Mohan-Zhang-u) [reply quarticgym/envs/beerfmtenv.py:255]: sure!

File: quarticgym/envs/mabdownenv.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        if self.normalize:
+            self.observation_space = spaces.Box(low=-1, high=1, shape=(self.observation_dim,))
+            self.action_space = spaces.Box(low=-1, high=1, shape=(self.action_dim,))
+        else:
+            self.observation_space = spaces.Box(low=self.min_observations, high=self.max_observations,
+                                                shape=(self.observation_dim,))
+            self.action_space = spaces.Box(low=self.min_actions, high=self.max_actions, shape=(self.action_dim,))
+            
+    def reward_function_standard(self, previous_observation, action, current_observation, reward=None):
+        r = self.compute_recover_rate(current_observation)  # input shape (2+1950,)
```
Comment (Sile-Tao) [inline quarticgym/envs/mabdownenv.py:83]: return self.compute_recover_rate(current_observation)  # input shape (2+1950,)

File: quarticgym/envs/mabdownenv.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        multi_states = self.previous_observation[2:] # (self.num_z_cap_load*13,) = 1950
+        self.downstream = DownModelHelper(num_z_cap_load=self.num_z_cap_load,
+            num_z_cap_elu=self.num_z_cap_elu,
+            num_z_cex_load=self.num_z_cex_load,
+            num_z_cex_elu=self.num_z_cex_elu,
+            num_z_loop=self.num_z_loop,
+            num_z_aex=self.num_z_aex,
+            num_r=self.num_r,
+            num_sim=self.num_sim_d,
+            delta_t=self.dt_d,
```
Comment (Sile-Tao) [inline quarticgym/envs/mabdownenv.py:129]: What does "_d" in num_sim_d and dt_d mean?

File: quarticgym/envs/reactorenv.py [hunk 13]
```diff
@@ -32,6 +34,134 @@
...
+    def build_controller(self):
+        
+        # stage and terminal cost in casadi symbolic form
+        l = mpc.getCasadiFunc(self.lfunc, [self.Nx,self.Nu], ["x","u"], funcname="l")
+        Pf = mpc.getCasadiFunc(self.Pffunc, [self.Nx], ["x"], funcname="Pf")
+        
+        # Create casadi function/model to be used in the controller
+        ode_casadi = mpc.getCasadiFunc(self.model.ode, [self.Nx, self.Nu], ["x","u"], funcname="odef")
+
+        # Set c to a value greater than one to use collocation
```
Comment (Sile-Tao) [inline quarticgym/envs/reactorenv.py:106]: shall we make c an argument to build_controller and set default value to 3?

File: quarticgym/envs/reactorenv.py [hunk 14]
```diff
@@ -32,6 +34,134 @@
...
+        contargs = dict(
+                N = {"t":self.Nt, "x":self.Nx, "u":self.Nu, "c":3},
+                verbosity=0,    # set verbosity to a number > 0 for debugging purposes otherwise use 0.
```
Comment (Sile-Tao) [inline quarticgym/envs/reactorenv.py:109]: Same here. Instead of writing many comments, we can write docstrings and make some of them function arguments.

**Review by Sile-Tao — COMMENTED** [review]
More comments left.

File: quarticgym/envs/mabdownenv.py [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+                                                shape=(self.observation_dim,))
+            self.action_space = spaces.Box(low=self.min_actions, high=self.max_actions, shape=(self.action_dim,))
+            
+    def reward_function_standard(self, previous_observation, action, current_observation, reward=None):
+        r = self.compute_recover_rate(current_observation)  # input shape (2+1950,)
+        return r
+    
+    def compute_recover_rate(self, x):  # Recovery yield = purified product/harvested mAb
+        harvested = x[-3]*x[-2]*x[-1]
+        bottom = np.max([20000*self.delta_t, x[-1]])
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/mabdownenv.py:88]: sorry this file is still under heavy development and everything is not finalized.. The shape, variable names and file names will be actively change and I will change according to your comments when everything is settled

[the remark this answers is not in the document]
Comment (Sile-Tao) [reply quarticgym/envs/mabdownenv.py:88]: I see. No worries

File: quarticgym/envs/mabdownenv.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+            observation = initial_state
+            self.init_observation = initial_state
+        else:
+            observation = self.sample_initial_state()
+            self.init_observation = observation
+        self.previous_observation = observation
+
+        # TOMODIFY: reset your environment here.
+        flow_from_upper = self.previous_observation[:2]
+        multi_states = self.previous_observation[2:] # (self.num_z_cap_load*13,) = 1950
```
[the remark this answers is not in the document]
Comment (Mohan-Zhang-u) [reply quarticgym/envs/mabdownenv.py:120]: the shapes are not finalized and will be changed.

**Review by xiaozhouwang — APPROVED** [review]

**Review by Sile-Tao — APPROVED** [review]

**Pull request closed without merging**